packages/lesswrong/lib/collections/sequences/helpers.js

```diff
@@ -1,17 +1,78 @@
 import Sequences from './collection.js';
 import { Posts } from '../posts/index.js';
 import Chapters from '../chapters/collection.js';
-
 import { Utils } from 'meteor/vulcan:core';
+import keyBy from 'lodash/keyBy';
+
+// TODO: Make these functions able to use loaders for caching.
 
 Sequences.getPageUrl = function(sequence, isAbsolute = false){
   const prefix = isAbsolute ? Utils.getSiteUrl().slice(0,-1) : '';
 
   return `${prefix}/s/${sequence._id}`;
 };
 
+Sequences.getAllPostIDs = async (sequenceId) => {
+  const chapters = await Chapters.find({sequenceId: sequenceId}, {sort: {number: 1}}).fetch()
+  let allPostIds = _.flatten(_.pluck(chapters, 'postIds'))
+  return allPostIds;
+}
+
 Sequences.getAllPosts = async (sequenceId) => {
-  const chapters = await Chapters.find({sequenceId: sequenceId}).fetch()
-  let allPostsIds = _.flatten(_.pluck(chapters, 'postIds'))
-  return await Posts.find({_id:{$in:allPostsIds}}).fetch()
-}
\ No newline at end of file
+  // Get the set of post IDs in the sequence (by joining against the Chapters
+  // table), sorted in reading order
+  const allPostIds = await Sequences.getAllPostIDs(sequenceId);
+  
+  // Retrieve those posts
+  const posts = await Posts.find({_id:{$in:allPostIds}}).fetch()
+  
+  // Sort the posts retrieved back into reading order and return them
+  const postsById = keyBy(posts, post=>post._id);
+  return _.map(allPostIds, id=>postsById[id]);
+}
+
+// Given a post ID and the ID of a sequence which contains that post, return the
+// next post in the sequence, or null if it was the last post. Does not handle
+// cross-sequence boundaries. If the given post ID is not in the sequence,
+// returns null.
+Sequences.getNextPostID = async function(sequenceId, postId) {
```

Comment:
Maybe factor out commonalities between the two functions? Not super important, but seems better for maintainability.

Reply:
By line count, most of this is bounds checks, which are different. It's possible to factor out a common function, but it would have to use a `delta` parameter, which is a more confusing pattern than this particular function pair is worth.